quality_of_service_demo/src/common_nodes.cpp

```diff
@@ -0,0 +1,100 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <chrono>
+#include <string>
+
+#include "quality_of_service_demo/common_nodes.hpp"
+
+using namespace std::chrono_literals;
+
+Talker::Talker(
+  const std::string & topic_name,
+  rclcpp::PublisherOptions<> pub_options,
+  size_t max_count,
+  std::chrono::milliseconds assert_node_period,
+  std::chrono::milliseconds assert_topic_period)
+: Node("talker"),
+  max_count_(max_count)
+{
+  RCLCPP_INFO(get_logger(), "Talker starting up");
+  publisher_ = create_publisher<std_msgs::msg::String>(topic_name, pub_options);
+  publish_timer_ = create_wall_timer(
+    500ms,
+    [this]() -> void {
+      std_msgs::msg::String msg;
+      msg.data = "Talker says " + std::to_string(count_++);
+      publisher_->publish(msg);
+      if (max_count_ > 0 && count_ > max_count_) {
+        publish_timer_->cancel();
+      }
+    });
+  // If enabled, create timer to assert liveliness at the node level
+  if (assert_node_period != 0ms) {
+    assert_node_timer_ = create_wall_timer(
+      assert_node_period,
+      [this]() -> void {
+        this->assert_liveliness();
+      });
+  }
+  // If enabled, create timer to assert liveliness on the topic
+  if (assert_topic_period != 0ms) {
+    assert_topic_timer_ = create_wall_timer(
+      assert_topic_period,
+      [this]() -> void {
+        publisher_->assert_liveliness();
+      });
+  }
+}
+
+void Talker::pause_for(std::chrono::milliseconds pause_length)
+{
+  if (pause_timer_) {
+    // Already paused - just ignore
```

Comment:
What if the `pause_length` is greater than the remaining time or even the original length of the previous pause?